casper/src/main/scala/coop/rchain/casper/blocks/merger/CasperDagMerger.scala

```diff
@@ -0,0 +1,180 @@
+package coop.rchain.casper.blocks.merger
+
+import cats.effect.Concurrent
+import cats.syntax.all._
+import coop.rchain.casper.PrettyPrinter
+import coop.rchain.casper.protocol.ProcessedDeploy
+import coop.rchain.casper.util.EventConverter
+import coop.rchain.casper.util.rholang.RuntimeManager
+import coop.rchain.casper.util.rholang.RuntimeManager.StateHash
+import coop.rchain.crypto.codec.Base16
+import coop.rchain.dag._
+import coop.rchain.rspace.Blake2b256Hash
+import coop.rchain.rspace.syntax._
+import coop.rchain.shared.{Log, Stopwatch}
+import coop.rchain.store.LazyKeyValueCache
+
+import scala.collection.Seq
+
+object CasperDagMerger {
+
+  /**
+    * Merge DAG.
+    * @param tips - top blocks of the DAG
+    * @param base - bottom blocks of the DAG, all tips should be connected with base
+    * @param dag - connection graph
+    * @return tuplespace state hash of the merged state and list of conflicting deploys to reject
+    */
+  def merge[F[_]: Concurrent: RuntimeManager: Log](
+      tips: Seq[MergingVertex],
+      base: MergingVertex,
+      dag: DagReader[F, MergingVertex]
+  ): F[(StateHash, Seq[ProcessedDeploy])] = {
+
+    // ordering is required for computing branches procedure to be replayable
+    implicit val ord: Ordering[MergingVertex] = (x: MergingVertex, y: MergingVertex) =>
+      Base16.encode(x.stateHash.toByteArray).compare(Base16.encode(y.stateHash.toByteArray))
+
+    for {
+      baseStateReader <- RuntimeManager[F].getHistoryRepo
+                          .reset(Blake2b256Hash.fromByteString(base.stateHash))
+      branchesSorted <- DagOps.computeSortedBranches(tips, base, dag)
+      // TODO extract syntax for streams to compute in parallel but preserve order
+      branchesIndexed <- Stopwatch.time(Log[F].info(_))(
```

Comment:
@tgrospic Is `Stopwatch` still a temporary solution or for long term now?
Why not Span?

Reply:
I am ok for long term usage.